File: sdk/python/examples/langgraph/22_human_in_the_loop.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Human-in-the-Loop — interrupt/resume workflow for human approval.

Demonstrates:
    - Using interrupt_before to pause at a node for human review
    - Resuming a graph with a MemorySaver checkpointer after human feedback
    - Practical use case: draft an email, wait for human approval, then send

    NOTE: This example simulates the human approval step locally.
          In production the graph would be stored by Agentspan between turns.

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from typing import TypedDict

from langchain_core.messages import HumanMessage, SystemMessage
from langchain_openai import ChatOpenAI
from langgraph.graph import StateGraph, START, END
from langgraph.checkpoint.memory import MemorySaver
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


class EmailState(TypedDict):
    request: str
    draft: str
    human_feedback: str
    final_email: str


def draft_email(state: EmailState) -> EmailState:
    response = llm.invoke([
        SystemMessage(content="You are a professional email writer. Draft a concise, polite email."),
        HumanMessage(content=f"Request: {state['request']}"),
    ])
    return {"draft": response.content.strip()}


def human_review(state: EmailState) -> EmailState:
    """This node is interrupted before execution so a human can provide feedback."""
    feedback = state.get("human_feedback", "")
    if not feedback or feedback.lower() == "approve":
        return {"final_email": state["draft"]}
    return {}


def revise_email(state: EmailState) -> EmailState:
    if not state.get("human_feedback"):
        return {"final_email": state["draft"]}
    response = llm.invoke([
        SystemMessage(content="Revise the email draft based on the feedback provided."),
        HumanMessage(
            content=f"Original draft:\n{state['draft']}\n\nFeedback:\n{state['human_feedback']}"
        ),
    ])
    return {"final_email": response.content.strip()}


def should_revise(state: EmailState) -> str:
    feedback = state.get("human_feedback", "")
    if feedback and feedback.lower() not in ("approve", "ok", "looks good"):
        return "revise"
    return "done"


builder = StateGraph(EmailState)
builder.add_node("draft", draft_email)
builder.add_node("human_review", human_review)
builder.add_node("revise", revise_email)

builder.add_edge(START, "draft")
builder.add_edge("draft", "human_review")
builder.add_conditional_edges("human_review", should_revise, {"revise": "revise", "done": END})
builder.add_edge("revise", END)

# MemorySaver checkpointer enables interrupt/resume between turns
checkpointer = MemorySaver()
graph = builder.compile(
    name="email_hitl_agent",
    checkpointer=checkpointer,
    interrupt_before=["human_review"],
)

if __name__ == "__main__":
    thread_config = {"configurable": {"thread_id": "email-thread-1"}}

    # ── Turn 1: draft the email ───────────────────────────────────────────────
    print("=== Turn 1: drafting email ===")
    for mode, chunk in graph.stream(
        {"request": "Schedule a team meeting for next Monday at 10am to discuss Q3 plans."},
        thread_config,
        stream_mode=["updates", "values"],
    ):
        if mode == "updates":
            for node, updates in chunk.items():
                if "draft" in updates:
                    print(f"\nDraft:\n{updates['draft']}\n")

    # ── Simulate human approval ───────────────────────────────────────────────
    print("=== Turn 2: human approves ===")
    for mode, chunk in graph.stream(
        {"human_feedback": "approve"},
        thread_config,
        stream_mode=["updates", "values"],
    ):
        if mode == "values" and chunk.get("final_email"):
            print(f"Final email:\n{chunk['final_email']}")

File: sdk/typescript/examples/langgraph/22-human-in-the-loop.ts
/**
 * Human-in-the-Loop -- real human approval gate within a LangGraph workflow.
 *
 * Demonstrates:
 *   - Draft -> Human Review -> Approve/Revise conditional workflow
 *   - A Conductor HUMAN task that pauses execution for actual human input
 *   - The human provides a verdict (APPROVE/REVISE) and feedback
 *   - Conditional routing based on human verdict
 *   - Running the full workflow through Agentspan via runtime.run()
 *
 * The workflow pauses at the review step and waits for a human to approve or
 * reject the draft via the Agentspan UI or API.
 *
 * In production you would use:
 *   import { StateGraph, START, END } from '@langchain/langgraph';
 *   import { humanTask } from '@agentspan/sdk/frameworks/langgraph';
 */

import { AgentRuntime } from '../../src/index.js';

// ---------------------------------------------------------------------------
// State type
// ---------------------------------------------------------------------------
interface EmailState {
  request: string;
  draft: string;
  review_verdict: string;
  review_feedback: string;
  final_email: string;
}

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function draftEmail(state: EmailState): Partial<EmailState> {
  // In production, this would call an LLM
  const draft =
    `Subject: Team Meeting - Monday 10am, Q3 Plans\n\n` +
    `Hi Team,\n\n` +
    `I'd like to schedule a team meeting for next Monday at 10:00 AM to discuss ` +
    `our Q3 plans and priorities. Please review the attached agenda beforehand.\n\n` +
    `Key topics:\n` +
    `- Q2 retrospective highlights\n` +
    `- Q3 OKRs and milestones\n` +
    `- Resource allocation\n\n` +
    `Please confirm your availability by end of day Friday.\n\n` +
    `Best regards,\n` +
    `[Your Name]`;
  return { draft };
}

function routeAfterReview(state: EmailState): string {
  if (state.review_verdict.toUpperCase() === 'APPROVE') {
    return 'finalize';
  }
  return 'revise';
}

function finalize(state: EmailState): Partial<EmailState> {
  return { final_email: state.draft };
}

function reviseEmail(state: EmailState): Partial<EmailState> {
  // In production, this would call an LLM with the feedback
  const revised =
    `Subject: Team Meeting - Monday 10am, Q3 Plans (Updated)\n\n` +
    `Hi Team,\n\n` +
    `Following reviewer feedback: "${state.review_feedback}"\n\n` +
    `I've updated the meeting invitation for next Monday at 10:00 AM ` +
    `to discuss our Q3 plans.\n\n` +
    `[Revised content addressing feedback]\n\n` +
    `Best regards,\n` +
    `[Your Name]`;
  return { final_email: revised };
}

// ---------------------------------------------------------------------------
// Mock compiled graph with HUMAN task node
// ---------------------------------------------------------------------------
const graph = {
  name: 'email_hitl_agent',

  invoke: async (input: Record<string, unknown>) => {
    const request = (input.input as string) ?? '';
    let state: EmailState = {
      request,
      draft: '',
      review_verdict: '',
      review_feedback: '',
      final_email: '',
    };

    // Step 1: Draft
    state = { ...state, ...draftEmail(state) };

    // Step 2: Simulate human review (in production, this pauses for real human input)
    // The Conductor HUMAN task would collect review_verdict and review_feedback
    state.review_verdict = 'APPROVE';
    state.review_feedback = 'Looks good!';

    // Step 3: Route based on verdict
    if (routeAfterReview(state) === 'finalize') {
      state = { ...state, ...finalize(state) };
    } else {
      state = { ...state, ...reviseEmail(state) };
    }

    return { output: state.final_email };
  },

  getGraph: () => ({
    nodes: new Map([
      ['__start__', {}],
      ['draft', {}],
      ['review', { taskType: 'HUMAN' }], // HUMAN task marker
      ['finalize', {}],
      ['revise', {}],
      ['__end__', {}],
    ]),
    edges: [
      ['__start__', 'draft'],
      ['draft', 'review'],
      // Conditional: review -> finalize | revise
      ['finalize', '__end__'],
      ['revise', '__end__'],
    ],
  }),

  nodes: new Map([
    ['draft', {}],
    ['review', { taskType: 'HUMAN' }],
    ['finalize', {}],
    ['revise', {}],
  ]),

  stream: async function* (input: Record<string, unknown>) {
    const request = (input.input as string) ?? '';
    let state: EmailState = {
      request,
      draft: '',
      review_verdict: '',
      review_feedback: '',
      final_email: '',
    };

    state = { ...state, ...draftEmail(state) };
    yield ['updates', { draft: { draft: state.draft } }];

    // Simulate human review (in production, execution pauses here)
    state.review_verdict = 'APPROVE';
    state.review_feedback = 'Looks good!';
    yield [
      'updates',
      {
        review: {
          review_verdict: state.review_verdict,
          review_feedback: state.review_feedback,
        },
      },
    ];

    if (routeAfterReview(state) === 'finalize') {
      state = { ...state, ...finalize(state) };
      yield ['updates', { finalize: { final_email: state.final_email } }];
    } else {
      state = { ...state, ...reviseEmail(state) };
      yield ['updates', { revise: { final_email: state.final_email } }];
    }

    yield ['values', { output: state.final_email }];
  },
};

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    const result = await runtime.run(
      graph,
      'Schedule a team meeting for next Monday at 10am to discuss Q3 plans.',
    );
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

main().catch(console.error);
